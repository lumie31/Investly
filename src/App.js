import React, { useState } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import {
  Container,
  Box,
  TextField,
  Grid,
  Paper,
  ButtonBase,
} from "@material-ui/core";
import Dialog from "@material-ui/core/Dialog";
import Slide from "@material-ui/core/Slide";

import Navbar from "./components/navbar";
import currentLoans from "./current_loans.json";

const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});

const useStyles = makeStyles((theme) => ({
  root: {
    margin: "20px auto",
  },
  card: {
    paddingLeft: "10px",
    paddingRight: "10px",
  },
  button: {
    backgroundColor: "#FFEA00",
    color: "#000",
    padding: theme.spacing(1.5, 4),
    border: "1px solid #FDE800",
    borderRadius: "4px",
    textTransform: "uppercase",
    fontWeight: "500",
    fontSize: "14px",
  },
  paper: {
    backgroundColor: "#F5F5F5",
  }
}));

// function to parse string into number
const parseNumber = (num) => {
  return parseFloat(num.split(",").join(""));
};

const initialState = currentLoans.loans.map((loan) => {
  return {
    ...loan,
    available: parseNumber(loan.available),
    amount: parseNumber(loan.amount),
    term_remaining: parseNumber(loan.term_remaining),
  };
});


export default function App() {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const [loans, setLoans] = useState(initialState);
  const [loanInput, setLoanInput] = useState("");
  const [selectedLoan, setSelectedLoan] = useState(null);
  const [investments, setInvestments] = useState([]);
  const [error, setError] = useState(null);

  const totalAvailable = initialState.reduce(
      (acc, curr) => (acc + curr.available), 0);

  const handleClickOpen = (loan) => {
    setOpen(true);
    setSelectedLoan(loan);
  };

  const handleClose = () => {
    setOpen(false);
    setSelectedLoan(null);
    setError(null);
    setLoanInput("");
  };

  const investAmount = () => {
    if (!loanInput || loanInput === "" || loanInput <= 0) {
      setError("Input cannot be empty");
      return;
    }
    if (parseNumber(loanInput) > selectedLoan.available) {
      setError("Your investment cannot be greater than your available amount");
      return;
    }

    let newInvestments = [...investments];
    const investmentIndex = newInvestments.findIndex(investment => investment.id === selectedLoan.id);

    let newLoans = [...loans];
    const loanIndex = newLoans.findIndex(loan => loan.id === selectedLoan.id);

    newLoans[loanIndex]["available"] = newLoans[loanIndex]["available"] - parseNumber(loanInput);
    newLoans[loanIndex]["showInvestedBadge"] = true

    if (investmentIndex !== -1) {
      newInvestments[investmentIndex]["invested"] = newInvestments[investmentIndex]["invested"] + parseNumber(loanInput);
    } else {
      newInvestments.push({
        id: selectedLoan.id,
        invested: parseNumber(loanInput)
      })
    }
    setLoans(newLoans);
    setInvestments(newInvestments);
    handleClose();
  }

  return (
    <Box>
      <Navbar />
      <Container maxWidth="sm" className={classes.root}>
        <Paper elevation={2} className={classes.paper}>
          <Box p={2}>
            <h1>Current Loans</h1>
            <Box>
              {loans.map((loan) => {
                return (
                  <Box key={loan.id} my={2} border="2px solid #F5F5F5">
                    <Card className={classes.card}>
                      <Grid container alignItems="center">
                        <Grid item xs={9}>
                          <h3>{loan.title}</h3>
                        </Grid>
                        {loan.showInvestedBadge && (
                          <Grid item xs={3}>
                            <Box component="span" color="green">
                              Invested
                            </Box>
                          </Grid>
                        )}
                        <Grid
                          item
                          xs={7}
                          sm={9}
                          md={9}
                          className={classes.grid}
                        >
                          <p>
                            <strong>Tranche:</strong> {loan.tranche}
                          </p>
                          <p>
                            <strong>Amount:</strong> £
                            {loan.amount.toLocaleString()}
                          </p>
                          <p>
                            <strong>Available:</strong> £
                            {loan.available.toLocaleString()}
                          </p>
                        </Grid>
                        <Grid item xs={5} sm={3} md={3}>
                          <ButtonBase
                            onClick={() => handleClickOpen(loan)}
                            className={classes.btnBase}
                            data-testid="invest"
                          >
                            <Box className={classes.button}>Invest</Box>
                          </ButtonBase>
                        </Grid>
                      </Grid>
                    </Card>
                  </Box>
                );
              })}
            </Box>
            <Box component="p" textAlign="center" p={0} m={0}>
              Total amount available for investments:
              <strong> £{totalAvailable.toLocaleString()}</strong>
            </Box>
          </Box>
        </Paper>

        <Dialog
          open={open}
          TransitionComponent={Transition}
          keepMounted
          onClose={handleClose}
          aria-labelledby="alert-dialog-slide-title"
          aria-describedby="alert-dialog-slide-description"
        >
          <Box width={{ md: "30vw" }} p={(2, 4)}>
            <Box component="h3" mb={0.5}>
              Invest in Loan
            </Box>
            <Box component="p" mt="0">
              {selectedLoan?.title}
            </Box>
            <Box>
              <p>
                <strong>Amount available:</strong> £
                {selectedLoan?.available.toLocaleString()}
              </p>
              <p>
                <strong>Loan ends in:</strong> {selectedLoan?.term_remaining}ms
              </p>
            </Box>
            <Grid container>
              <Grid item xs={12}>
                <p>Investment amount (£)</p>
              </Grid>
              <Grid item xs={6} md={8}>
                <TextField
                  variant="outlined"
                  label="Amount"
                  type="number"
                  size="small"
                  value={loanInput.toLocaleString()}
                  onChange={(e) => setLoanInput(e.target.value)}
                  fullWidth
                />
              </Grid>
              &nbsp;
              <Grid item xs={4} md={3}>
                <ButtonBase 
                onClick={investAmount}
                >
                  <Box className={classes.button}>invest</Box>
                </ButtonBase>
              </Grid>
              <Grid item xs={6} md={8}>
                {error && (
                  <Box color="red" fontSize="12px" mx="auto" component="span">
                    {error}
                  </Box>
                )}
              </Grid>
            </Grid>
          </Box>
        </Dialog>
      </Container>
    </Box>
  );
}
