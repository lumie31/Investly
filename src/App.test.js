import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import App from './App';

afterEach(cleanup);
test('should take a snapshot', () => {
    const { asFragment } = render(<App />)
    expect(asFragment(<App />)).toMatchSnapshot()
})

test('should render text when app loads', () => {
  const { getByText } = render(<App />);
  const linkElement = screen.getByText(/investly/i);
  expect(linkElement).toBeInTheDocument();
  expect(getByText("Current Loans")).toBeInTheDocument();
});

test("button should be enabled when app loads", () => {
  const { getAllByTestId } = render(<App />);
  expect(getAllByTestId("invest")[0]).not.toHaveAttribute("disabled");
});

test("does not expect to haved called onClick prop when app loads", () => {
  const handleClick = jest.fn();
  render(<App/>);
  fireEvent.click(screen.getAllByText(/invest/i)[0]);
  expect(handleClick).toHaveBeenCalledTimes(0);
});