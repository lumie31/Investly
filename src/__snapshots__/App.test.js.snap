// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5"
  >
    <div
      class="makeStyles-root-6"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h6
            class="MuiTypography-root makeStyles-title-8 MuiTypography-h6"
          >
            INVESTLY
          </h6>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              ...mortgages simplified
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
    <div
      class="MuiContainer-root makeStyles-root-1 MuiContainer-maxWidthSm"
    >
      <div
        class="MuiPaper-root makeStyles-paper-4 MuiPaper-elevation2 MuiPaper-rounded"
      >
        <div
          class="MuiBox-root MuiBox-root-9"
        >
          <h1>
            Current Loans
          </h1>
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <div
              class="MuiBox-root MuiBox-root-11"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
                  >
                    <h3>
                      Voluptate et sed tempora qui quisquam.
                    </h3>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 MuiGrid-grid-sm-9 MuiGrid-grid-md-9"
                  >
                    <p>
                      <strong>
                        Tranche:
                      </strong>
                       A
                    </p>
                    <p>
                      <strong>
                        Amount:
                      </strong>
                       £85,754
                    </p>
                    <p>
                      <strong>
                        Available:
                      </strong>
                       £11,959
                    </p>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-3 MuiGrid-grid-md-3"
                  >
                    <button
                      class="MuiButtonBase-root"
                      data-testid="invest"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-12 makeStyles-button-3"
                      >
                        Invest
                      </div>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-13"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
                  >
                    <h3>
                      Consectetur ipsam qui magnam minus dolore ut fugit.
                    </h3>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 MuiGrid-grid-sm-9 MuiGrid-grid-md-9"
                  >
                    <p>
                      <strong>
                        Tranche:
                      </strong>
                       B
                    </p>
                    <p>
                      <strong>
                        Amount:
                      </strong>
                       £85,754
                    </p>
                    <p>
                      <strong>
                        Available:
                      </strong>
                       £31,405
                    </p>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-3 MuiGrid-grid-md-3"
                  >
                    <button
                      class="MuiButtonBase-root"
                      data-testid="invest"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-14 makeStyles-button-3"
                      >
                        Invest
                      </div>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-15"
            >
              <div
                class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
                  >
                    <h3>
                      Dolores repudiandae ut voluptas unde laborum quaerat et sapiente.
                    </h3>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 MuiGrid-grid-sm-9 MuiGrid-grid-md-9"
                  >
                    <p>
                      <strong>
                        Tranche:
                      </strong>
                       C
                    </p>
                    <p>
                      <strong>
                        Amount:
                      </strong>
                       £85,754
                    </p>
                    <p>
                      <strong>
                        Available:
                      </strong>
                       £12,359
                    </p>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-sm-3 MuiGrid-grid-md-3"
                  >
                    <button
                      class="MuiButtonBase-root"
                      data-testid="invest"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-16 makeStyles-button-3"
                      >
                        Invest
                      </div>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <p
            class="MuiBox-root MuiBox-root-17"
          >
            Total amount available for investments:
            <strong>
               £55,723
            </strong>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
